### PR TITLE
implemented MarkerStyle.register and relative tests

### LINKED ISSUE #29047: [ENH]: Registering custom markers
### Problem

While working on a library to make styles (with custom colors, etc...) I discovered that there is no easy way to register custom markers, unlike for colors and the like.
I found a workaround digging in `markers.py`:
```python
from matplotlib.markers import MarkerStyle
...
MarkerStyle.markers[marker_name] = marker_name
setattr(MarkerStyle, f'_set_{marker_name}', lambda self, path=marker_path: self._set_custom_marker(path))
```
which seems to work, and allows to use the new maker in other files as
```python
plt.plot(x, y, marker = marker_name)
```
However, this code is quite clumsy and inelegant!

### Proposed solution

It would be nice to have a way to specify
```python
MarkerStyle.register(marker_name, marker_path)
```
similarly to [how it is done for colormaps](https://matplotlib.org/stable/api/cm_api.html). 

This would be pretty easy because it could leverage internally `MarkerStyle._set_custom_marker`, which already implements most of the necessary functionality!

If this is welcome, I would be happy to have a go and submit a PR!
I have found that this is quite nice to drive up the engagement of students to be able to easily play with visuals in this way :)

### PR DESCRIPTION
**Pull request by LorenzoPeri17**
## PR summary

Implements `MarkerStyle.register` to globally register a custom marker, specifying a name and a `Path`, and adds testing for the new functionality.
Closes #29047 

I wasn't sure if the change deserved a change in the docs (maybe to the [marker reference](https://matplotlib.org/stable/gallery/lines_bars_and_markers/marker_reference.html) example?). If it does I'm happy to give it a stab and add it to the PR!

## PR checklist

- [X] "closes #0000" is in the body of the PR description to [link the related issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)
- [X] new and changed code is [tested](https://matplotlib.org/devdocs/devel/testing.html)
- [?] *Plotting related* features are demonstrated in an [example](https://matplotlib.org/devdocs/devel/document.html#write-examples-and-tutorials)
- [?] *New Features* and *API Changes* are noted with a [directive and release note](https://matplotlib.org/devdocs/devel/api_changes.html#announce-changes-deprecations-and-new-features)
- [X] Documentation complies with [general](https://matplotlib.org/devdocs/devel/document.html#write-rest-pages) and [docstring](https://matplotlib.org/devdocs/devel/document.html#write-docstrings) guidelines

### CODE REVIEW

**Review by jklymak — COMMENTED** [review]
I'm not sure how this is meant to work.  There doesn't seem to be any way to query this registry, and no way to remove elements?  Do we want this to be closer to colormap registry?

**Comment by LorenzoPeri17** [comment]
Thanks for the comment! While looking at the code for the existing `MarkerStyle` I saw that default named markers are implemented via `getattr(self, '_set_' + self.markers[marker])`, so I figured that this would be the simplest way to extend the code to custom markers.

To have something closer to the `ColormapRegistry`, I could either implement and `unregister` method wrapping `delattr` (querying after registering can already be acheved with `MarkerStyle("custom_marker").get_path()`), or maybe I could create a new mapping `MarkerPathRegistry` where `MarkerStyle` looks for the `Path` if a name is given, rather than using `getattr`? In the latter case, maybe also the `_set_[marker name]` methods should belong to the registry?

Happy to hear any feedback! :)

**Comment by timhoffm** [comment]
Thanks for your willingness to contribute. I'm afraid `MarkerStyle.register` is not appropriate. While somewhat similar, the design pattern of `MarkerStyle` is substantially different from Colormap (or ColorSequence).

**Instance specification**
- `Colormap` takes only the data to create the mapping. (It also has a name, but that's more a label for information)
  In that sense `Colormap` is a content-agnostic data container - you can create arbitrary colormaps, but the class does not know anything about conrecte instances.
- In contrast, `MarkerStyle` is much fatter. It has the philosophy to accept anything that suits as a marker definition: predefined mnemonic strings, symbols as mathtext, arbitrary paths, another marker (plus modifications). Consequently, there's much more logic behind the scenes. As a result of this concept, we can simply accept any "marker definition" for `marker` kwargs - internally this is feed to the constructor, we don't need a registry to look up a name.

**Identifiers**
Colormaps have names, markers have a 1-letter mnemonic. The 1-letter chars are naturally limited - making them freely registerable has limited benefit and there's a high chance of name clashes. While one could let the user add longer marker names, this is a deviation from the current concept and e.g. these longer names could not be used e.g. in `plot` *fmt* strings.

**Registry**
`ColormapRegistry` is a dedicated concept. The main purpose is to support `cmap=name` with a string name usage. The registry handles all infrastructure aspects such as preventing overwriting a registered colormap (reassigning a name), prevent the modification of builtin colormaps etc.
A marker style registry would need to care for these as well.

Overall, `MarkerStyle` is currently scoped differently than colormaps/colorsequences. It's not straight forward to create a registry for it. I'd be reluctant to have different approaches to registering (dedicated ColormapRegistry vs. `MarkerStyle.register`), and I certainly don't want to dynamically monkey-patch additional functions onto `MarkerStyle`. While likely possible, I believe adding marker registration would require thorough investigation and likely some rearchitecting of marker handling. I'm not clear it's worth the effort.
In contrast, I'd ask: Is there a substrantial benefit in providing a registry? / What kind of markers would you add? - I assume most people can live with the existing markers: the builtin mnemonics and mathtext markers already have a short string represenation and are ample. Creating markers from other Paths should only needed for very specific and rare cases, and then passing a path should be good enough. 
From my perspective, the next-helpful addition would be a shortcut for creating a marker from an arbitrary glyph, which currently needs and explicit `TextPath` conversion (see e.g. https://matplotlib.org/devdocs/gallery/lines_bars_and_markers/multivariate_marker_plot.html).

**Comment by jklymak** [comment]
We have registries for colormaps and scales.  However, we do not have registries for norms, transforms, colors, linestyles, and markers.  

I'm struggling to understand the point of the registries - however, if there is one advantage I can think of, it would be that you can query the registry to see what is available, so if a library has registered its own colormap (etc) I would now see that in the list.  At the very least if we add a marker style registry, I think we should be able to query it.  As @timhoffm suggests, it should probably also have a warning if we try and re-use an existing marker's character.    

But overall, I'm not clear on the cost/benefit of adding a bunch more registries.

**Comment by LorenzoPeri17** [comment]
Thanks for the comments!

> What kind of markers would you add?

As @story645 was mentioning, I was also thinking on domain-specific markers (sports, sciences, company logos, etc..), or simply a `Path` created by parsing svgs. I find the latter particularly nice for teaching (fire and snowflake markers hit different than colors + a legend :)

To support that kind of usecase now the best solution I have found is to define a dictionary of either `Path` of `MarkerStyle` objects, to then use as `ax.plot(..., marker=marker_dict['custom_marker'])`, which is a little clunky and has sometimes annoying/unexpected edgecases. 

I appreciate that the code in this PR is a bit of a hack dynamically adding methods to the class, and that `MarkerStyle` is the class that both keeps hold of the `Path`s for the markers, as well as being responsible for much of the marker styling. 

> Colormaps have names, markers have a 1-letter mnemonic. The 1-letter chars are naturally limited - making them freely registerable has limited benefit and there's a high chance of name clashes. While one could let the user add longer marker names, this is a deviation from the current concept and e.g. these longer names could not be used e.g. in plot fmt strings.

Yes, I would agree that probably it would not be the best idea to allow users to define 1-char mnemonics, and that those could be reserved for default markers. But letting the user define longer-named markers with the understanding that those are only usable within the `marker` keyword seemed ok to me. Plus, it already mimics the [current usage of custom markers](https://matplotlib.org/stable/gallery/lines_bars_and_markers/marker_reference.html)

```python
ax.plot(..., marker=marker_path)
```

All it would solve is not having to keep a `Path` or `MarkerStyle` object around each time it needs to be used, which is more user-friendly. It would also help with setting a particular plot style. One could register a custom marker, and setting it as the default in `rcParams` (or even cycling automatically between custom markers). This cannot currently be done with always passing a path afaik.

> From my perspective, the next-helpful addition would be a shortcut for creating a marker from an arbitrary glyph, which currently needs and explicit TextPath conversion 

If this is helpful, I would be happy to address this in this PR. Maybe when passing a string to the `MarkerStyle` constructor if this is neither a known marker mnemonic nor a `mathtex`, the default behaviour could be to convert it to a `TextPath` (maybe with a warning?), rather than raising a `ValueError` as it does now?

> At the very least if we add a marker style registry, I think we should be able to query it. As @timhoffm suggests, it should probably also have a warning if we try and re-use an existing marker's character.

This could be relatively straightforward to implement within `MarkerStyle`, as it already keeps a record of all known markers in an internal dictionary.

Anyway, I would be happy to implement warnings/keep working on this PR if you think it is useful. Otherwise I'm also happy to close it if this functionality is out of scope :)

**Comment by timhoffm** [comment]
Hatch is a completely different story. The current hatches are limited to a fixed set of base shapes with some rules to combine and  repeat. There are fundamental limitations on what hatches you can use. With markers, you can do everything (using paths) and a lot of markers are *easily* available through strings (mnemonics plus maththext).

> > The default font may not have the desired glyph, and when short-cutting TextPath we loose the ability to use a different font.
> 
> Can there be a "force_not_textpath" flag or something so folks can opt out of the short cut?

I don't understand the motivation in this context. If you want to use a differnt font, the opt-out alone does not help, beause you have to pass the desired font somehow. Disregarding context, you could do that, but introducing special syntax for it is not worth it, given the only use case is "current mnemonic char as marker", which should be rare, and therefore the builtin opt-out for wrapping the mnemonic in a `TextPath` is sufficient.
Generally, though there are a number of API variants (favourably string prefixes e.g. `marker="c:+"` / `marker="c:😊"`) if we feel an automatism that some chars are interpreted as mnemonic and some are converted to text paths is too magic. But we should first decide whether the  fundamental idea "turn character into marker" is worth having a shortcut so that one does not need to explicitly create a TextPath.  I'm sure we can figure out a reasonable API.

**Comment by jklymak** [comment]
> While true, the motivation for a markerstyle registry is to more easily support/drop in third party markers - the whole point being that you don't have to make 'em b/c someone else did.

I don't see why a registry is better than the thrid-party providing a class?

**Comment by timhoffm** [comment]
> the motivation for a markerstyle registry is to more easily support/drop in third party markers

**TL;DR: a markerstyle registry is difficult to implement and would have limited benefit. IMHO the effort and API hassle is not worth the benefit.**



I feel, I'm repeating myself, but to make my points clear:

- Is there a substantial need for third-party makers? I claim no, because we already have lots of markers, and could even extend to arbitrary unicode symbols.
- Would a registry be a substantial benefit? I claim no. Needing to provide special markers to other people should be relatively rare (see the first bullet point). And for the few relevant cases, providing a module/namespace/dict with MarkerStyle instances is only slightly less convenient.
- Would it be easy to implement a marker registry? Again no (https://github.com/matplotlib/matplotlib/pull/29064#issuecomment-2453644797), not when taking the existing MarkerStyle into account and striving for consistent APIs and user experience.

I can list several other topics, that I'd rather spend time on, because they'd have a significantly larger impact.

**Comment by timhoffm** [comment]
> Frankly, I would very much appreciate a public technical roadmap so that there'd be some guidance, based on rough consensus, for evaluating technical proposals. My motivation here has been "mpl as building block library/let's make it easier for third parties"

A technical roadmap (beyond the RSE topics) is difficult because we don't have resources to feed that - everybody can work on what they like. What we have though, is a vision and strategy for the library - maybe that should be agreed upon and written out more explicitly.
My take: First sentence from the docs: "Matplotlib is a comprehensive library for creating static, animated, and interactive visualizations."
- comprehensive: We aim to enable all kinds of visualiztions
- common "basic" plots should be easy - we provide convenient high-level functionality for that (e.g. `scatter()`, legends etc.)
- complex visualizations should be possible - we provide building blocks that allow to create these  
  we encourage to create wrappers (possibly released as 3rd party packages) around these building blocks to create high-level interfaces for (domain-)specific plots, because high-level interfaces are desirable but the landscape of possible visualizations is too large for us to implement and maintain all kinds of specific visualizations within Matplotlib.

All work on the library should be aligned with that. Whether a concrete idea should be incorporated needs individual judgement, and depends on required effort, benefit, complexity, consistency, maintanance cost etc. I don't think we can give more concrete guidance with an abstract description. This needs case-by-case discussion - as we've done here.

To give some concrete ideas, here's an incomplete list of topics that I find worth advancing (most of them already have issues or PRs discussing them):
- grouped bar plots (I consider them "common basic plots" and I'm always embarrased if I have to tell people they have to calculate bar positions manually.)
- rearchtecting configuration management (aka rcParams)
- rearchitecting tick management
- better decoupling of mixed "pyplot" concerns: backend/event loop management vs. the function-based plot API
- consolidation of examples
- API consistency improvements

**Comment by jklymak** [comment]
> Frankly, I would very much appreciate a public technical roadmap so that there'd be some guidance, based on rough consensus, for evaluating technical proposals.

I agree that roadmapping this would be almost impossible.  API additions should be weighed on a cost/benefit basis.  

For this proposal I see the benefits as allowing a short form for the marker, and the possibility of querying all "registered" markers.  I think both of these "benefits" are somewhat reduced in modern programming practice by the ability of IDEs to surface properties and classes.  

In this case, the costs are API complexity and maintainability.  I'd also argue it is confusing when other libraries implicitly register things in our registry when their library is imported.  Panda's automagically adding datetime converter to our units registry causes a lot of confusion, because it is not obvious to downstream users that this has happened at all.  You could easily see someone trying to do `plot(x, y, marker='b')` where `b` is mplsoccer's marker for a ball, and someone trying to use that snippet without importing mplsoccer, and not understanding why it throws an error.  `plot(x, y, marker=mplsoccer.markers.soccerball)` is clear, and easy to type in most IDEs. 

But that is just my opinion - I'm pretty skeptical of the registries overall, and very skeptical of adding new ones.

**Comment by timhoffm** [comment]
> a contributor who is willing to spend their time on something

Note: Unless the topic is really straight forward, this means a core developer has to spend an equal amount of time for guidance and review. Because we cannot easily scale core developer time, this is a pratical limiting factor and a decision criterion on the cost side. - In fact, this is deciding aspect in my cost-benefit judgement here. While I'm sure one could get a reasonable marker registry into matplotlib, I estimate I would have to invest several hours to shepherd that through (since it's heavily API-related I feel would need to be involved - but even if another core developer would guide the contribution the inversted time stays the same). I believe these hours would not be well spend for the (small) benefit of a marker registry.

> would be good to have a discussion/consensus on "what's the criteria for something getting a registry?"

I feel this is an academic question and can be answered more practically for concrete cases. I wouldn't want to go about and answer such types of questions as part of a strategy. But since you asked:
A registry makes sense for complex objects, that you cannot easily specify inline, and for which exists a reasonable likelyhood that the builtin variants are not suffficient. It allows for an up-front specification and referencing that specification later by name.

- Positive example: Colormaps totally make sense here - you don't want to specifiy dozens of colors each time.
- Negative example: A color registry is unneeded. Named colors stand by themselves and there are plenty. Other ways of color specifications are concise. If you want to reuse them, assigning a name (`fancywhite = 'fff8f8'; plot(..., color=fancywhite)` is good enough. No need for the extra concert of a registry.
- Negative example 2: [Join styles](https://matplotlib.org/stable/gallery/lines_bars_and_markers/joinstyle.html#sphx-glr-gallery-lines-bars-and-markers-joinstyle-py) would need a complex specification, but the likelyhood of needing something beyond the builtin join styles is small.

Marker is an edge case. Path-based specifications can complex. On a green-field, I'd likely go for a marker registry. But the way to get there from the current marker architecture is more complicated, and that tips the balance. (Note how the current state influences the argument, which is why I above recommended to discuss concrete cases, but not abstract theories up-front).

**Comment by timhoffm** [comment]
Based in https://github.com/matplotlib/matplotlib/pull/29064#issuecomment-2461009400 I propose to close this PR. Please speak up if you fundamentally disagree and have additional arguments.
